libndt7/src/main/java/io/ooni/mk/nuvolari/libndt7/Client.java

```diff
@@ -1,100 +1,155 @@
         // NOTHING
     }
 
-    public void onError(String error) {
+    public void onError(@Nullable String error) {
+        Log.d(TAG, "onError: " + error);
         // NOTHING
     }
 
-    public void onServerDownloadMeasurement(Measurement measurement) {
+    public void onServerDownloadMeasurement(@NonNull Measurement measurement) {
+        Log.d(TAG, "onServerDownloadMeasurement: " + measurement);
         // NOTHING
     }
 
-    public void onClientDownloadMeasurement(Measurement measurement){
+    public void onClientDownloadMeasurement(@NonNull Measurement measurement){
+        Log.d(TAG, "onClientDownloadMeasurement: " + measurement);
         // NOTHING
     }
 
     @Override
     public final void onOpen(WebSocket ws, Response resp) {
-        onLogInfo("WebSocket connection established");
+        onLogInfo("WebSocket onOpen");
     }
 
     @Override
-    public final void onMessage(WebSocket ws, String text) {
+    public final void onMessage(WebSocket ws,
+                                String text) {
+        onLogInfo("WebSocket onMessage");
         count += text.length();
         periodic();
 
         Measurement measurement = new Measurement();
-        JSONObject doc;
+
+        @NonNull
+        final JSONObject doc;
+
+        // Root
         try {
             doc = new JSONObject(text);
-            measurement.elapsed = doc.getDouble("elapsed");
-            measurement.numBytes = doc.getDouble("num_bytes");
-        } catch (JSONException exc) {
-            // TODO(bassosimone): how to handle this failure? This is a case
-            // where the JSON has unexpected fields.
+        } catch (JSONException e) {
+            Log.d(TAG, "could not parse root json", e);
             return;
         }
+
+        // Top level
         try {
-            JSONObject bbrInfo = doc.getJSONObject("bbr_info");
-            measurement.bbrInfo = new BBRInfo();
-            try {
-                measurement.bbrInfo.bandwidth = bbrInfo.getDouble("bandwidth");
-                measurement.bbrInfo.RTT = bbrInfo.getDouble("rtt");
-            } catch (JSONException exc) {
-                // TODO(bassosimone): how to handle this failure? Like above
-                // this is a case where the JSON has unexpected fields.
-                return;
-            }
-        } catch (JSONException exc) {
-            // NOTHING: bbrInfo is optional
+            measurement.elapsed = doc.getDouble("elapsed");
+        } catch (JSONException e) { }
+
+        // tcp_info
+        try {
+            JSONObject tcpInfoObject = doc.getJSONObject("tcp_info");
+            measurement.tcpInfo = new Measurement.TcpInfo();
+            measurement.tcpInfo.smoothedRtt = tcpInfoObject.getDouble("smoothed_rtt");
+            measurement.tcpInfo.rttVar = tcpInfoObject.getDouble("rtt_var");
+        } catch (JSONException e) {
+            Log.d(TAG, "did not find tcp_info", e);
+            return;
```

Comment:
So, `tcp_info` is actually optional (the only non optional field is `elapsed`)

Reply:
Gotcha.  See comment above, let's revamp the json handling